tests/unit/s2n_self_talk_certificates_test.c

```diff
@@ -0,0 +1,142 @@
+        EXPECT_OK(s2n_test_load_certificate(&chain, pem,
+                test_certs[cert_i][0], test_certs[cert_i][1]));
+
+        for (size_t version_i = 0; version_i < s2n_array_len(test_versions); version_i++) {
+            uint8_t version = test_versions[version_i];
+
+            /* We intentionally use the default policies.
+             * The default policies should support all certificate types.
+             */
+            const char *security_policy = "default";
+            if (version >= S2N_TLS13) {
+                security_policy = "default_tls13";
+            } else if (version < S2N_TLS12) {
+                /* The default policies don't support legacy versions */
+                security_policy = "test_all";
+            }
+
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain));
+            EXPECT_SUCCESS(s2n_config_add_pem_to_trust_store(config, pem));
+            EXPECT_SUCCESS(s2n_config_set_verify_host_callback(config, s2n_noop_verify_host_fn, NULL));
+            EXPECT_SUCCESS(s2n_config_set_max_blinding_delay(config, 0));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, security_policy));
+
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server);
+            EXPECT_SUCCESS(s2n_connection_set_config(server, config));
+            server->server_protocol_version = version;
+
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client);
+            EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+            client->client_protocol_version = version;
+
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client, server, &io_pair));
+
+            bool handshake_success =
+                    (s2n_negotiate_test_server_and_client(server, client) == S2N_SUCCESS);
+            if (handshake_success) {
+                EXPECT_EQUAL(server->actual_protocol_version, version);
+                EXPECT_EQUAL(client->actual_protocol_version, version);
+            }
+
+            const char *error_message = "Handshake failed";
+            if (!s2n_test_is_valid(chain, version)) {
+                error_message = "Handshake unexpectedly succeeded";
+            }
+            if (handshake_success != s2n_test_is_valid(chain, version)) {
+                fprintf(stderr, "%s version=%i cert=%s\n",
+                        error_message, version, test_certs[cert_i][0]);
+                FAIL_MSG(error_message);
+            }
```

Comment:
nit: store bool to reuse
```suggestion
            bool valid_test = s2n_test_is_valid(chain, version)
            if (!valid_test) {
                error_message = "Handshake unexpectedly succeeded";
            }
            if (handshake_success != valid_test) {
                fprintf(stderr, "%s version=%i cert=%s\n",
                        error_message, version, test_certs[cert_i][0]);
                FAIL_MSG(error_message);
            }
```

Reply:
In hindsight I'm not a fan of s2n_test_is_valid at all. I just put more info on each test case.